File: src/lib/quizValidation.ts
import { differenceInCalendarDays, format, parseISO } from "date-fns";

import type { AnswerValue, QuizQuestion } from "@/lib/quiz";

export const MAX_TRIP_NIGHTS = 14;
export const MAX_TRAVELERS_TOTAL = 20;
export const MIN_CITY_LENGTH = 2;
export const MAX_CITY_LENGTH = 120;

const CITY_PATTERN = /^[\p{L}\p{M}0-9\s,'.\-()]+$/u;

function normalizeCityToken(name: string): string {
  let token = name.trim().toLowerCase();
  if (token.startsWith("al-")) {
    token = token.slice(3);
  }
  return token;
}

export function cityComparableKey(label: string): string {
  const parts = label.split(",").map((part) => part.trim()).filter(Boolean);
  if (!parts.length) return "";
  if (parts.length >= 3 && parts[0].includes("(")) {
    return normalizeCityToken(parts[1]);
  }
  return normalizeCityToken(parts[0]);
}

export function citiesConflict(origin: string, destination: string): boolean {
  const originClean = origin.trim();
  const destClean = destination.trim();
  if (!originClean || !destClean) return false;

  const originKey = cityComparableKey(originClean);
  const destKey = cityComparableKey(destClean);
  if (originKey && destKey) return originKey === destKey;

  return originClean.toLowerCase() === destClean.toLowerCase();
}

export function isValidCityFormat(value: string): boolean {
  const trimmed = value.trim();
  return (
    trimmed.length >= MIN_CITY_LENGTH &&
    trimmed.length <= MAX_CITY_LENGTH &&
    CITY_PATTERN.test(trimmed)
  );
}

export function validateDateRange(value: { start: string; end: string }): string | null {
  if (!value.start || !value.end) {
    return "Select both departure and return dates.";
  }
  if (value.end < value.start) {
    return "Return date must be on or after departure.";
  }

  const today = format(new Date(), "yyyy-MM-dd");
  if (value.start < today) {
    return "Departure cannot be in the past.";
  }

  const nights = differenceInCalendarDays(parseISO(value.end), parseISO(value.start));
  if (nights > MAX_TRIP_NIGHTS) {
    return `Trips can be up to ${MAX_TRIP_NIGHTS} nights (${MAX_TRIP_NIGHTS + 1} days).`;
  }

  return null;
}

export function validateTravelers(value: { adults: number; children: number }): string | null {
  if (value.adults < 1) return "At least one adult is required.";
  if (value.children < 0) return "Children count cannot be negative.";
  if (value.adults > 20) return "Maximum 20 adults per trip.";
  if (value.children > 10) return "Maximum 10 children per trip.";
  if (value.adults + value.children > MAX_TRAVELERS_TOTAL) {
    return `Maximum ${MAX_TRAVELERS_TOTAL} travelers per trip.`;
  }
  return null;
}

export function validateChoice(question: QuizQuestion, value: AnswerValue | undefined): string | null {
  if (value === undefined) return "Please select an option.";

  const allowed = new Set(question.options.map((o) => o.option_key));

  if (question.multi) {
    if (!Array.isArray(value)) return "Invalid selection.";
    if (question.key !== "travel_extras" && value.length === 0) return "Select at least one option.";
    if (question.key === "theme" && value.length > 2) return "Pick up to 2 interests.";
    if (value.some((v) => !allowed.has(v))) return "Invalid selection.";
    return null;
  }

  if (typeof value !== "string" || !allowed.has(value)) return "Please select an option.";
  return null;
}

export function validateCityField(
  value: string | undefined,
  options: { confirmedFromList: boolean; searchSettled: boolean; hasSuggestions: boolean },
): string | null {
  const trimmed = (value ?? "").trim();
  if (!trimmed) return "Enter a city name.";

  if (!isValidCityFormat(trimmed)) {
    return "Use letters and basic punctuation only (2–120 characters).";
  }

  if (options.confirmedFromList) return null;

  if (!options.searchSettled) return null;

  if (options.hasSuggestions) {
    return "Pick a city from the suggestions for accurate results.";
  }

  if (trimmed.length < 3) {
    return "Enter at least 3 characters or pick from suggestions.";
  }

  return null;
}

export function validateQuestionStep(
  question: QuizQuestion,
  value: AnswerValue | undefined,
  cityMeta?: { confirmedFromList: boolean; searchSettled: boolean; hasSuggestions: boolean },
): string | null {
  if (value === undefined) return null;

  switch (question.question_type) {
    case "choice":
      return validateChoice(question, value);
    case "date_range": {
      const dates = value as { start: string; end: string };
      if (!dates.start || !dates.end) return null;
      return validateDateRange(dates);
    }
    case "travelers":
      return validateTravelers(value as { adults: number; children: number });
    case "text":
      if (question.key === "origin" || question.key === "destination") {
        return validateCityField(typeof value === "string" ? value : undefined, cityMeta ?? {
          confirmedFromList: false,
          searchSettled: false,
          hasSuggestions: false,
        });
      }
      if (typeof value !== "string" || value.trim().length < 2) {
        return "Enter at least 2 characters.";
      }
      return null;
    default:
      return null;
  }
}

export function isQuestionStepReady(
  question: QuizQuestion,
  value: AnswerValue | undefined,
  cityMeta?: { confirmedFromList: boolean; searchSettled: boolean; hasSuggestions: boolean },
): boolean {
  if (value === undefined) return false;

  switch (question.question_type) {
    case "choice":
      if (question.key === "travel_extras" && value === undefined) return true;
      return validateChoice(question, value) === null;
    case "date_range": {
      const dates = value as { start: string; end: string };
      if (!dates.start || !dates.end) return false;
      return validateDateRange(dates) === null;
    }
    case "travelers":
      return validateTravelers(value as { adults: number; children: number }) === null;
    case "text":
      if (question.key === "origin" || question.key === "destination") {
        if (typeof value !== "string" || !value.trim()) return false;
        return (
          validateCityField(value, cityMeta ?? {
            confirmedFromList: false,
            searchSettled: false,
            hasSuggestions: false,
          }) === null
        );
      }
      return typeof value === "string" && value.trim().length >= 2;
    default:
      return false;
  }
}

export function validateCrossFields(
  questions: QuizQuestion[],
  answers: Record<number, AnswerValue>,
): string | null {
  const byKey = new Map(questions.map((q) => [q.key, q]));
  const originQ = byKey.get("origin");
  const destQ = byKey.get("destination");
  const destKnownQ = byKey.get("destination_known");
  const includeFlightsQ = byKey.get("include_flights");

  const destKnown = destKnownQ ? answers[destKnownQ.id] : "yes";
  if (destKnown === "not_sure") return null;

  const wantsFlights = includeFlightsQ ? answers[includeFlightsQ.id] === "yes" : false;
  if (!wantsFlights) return null;

  if (!originQ || !destQ) return null;

  const origin = answers[originQ.id];
  const destination = answers[destQ.id];
  if (typeof origin !== "string" || typeof destination !== "string") return null;

  if (citiesConflict(origin, destination)) {
    return "Departure airport must serve a different city than your destination.";
  }

  return null;
}

File: backend/app/services/quiz_validation.py
"""Server-side validation for quiz and preferences submissions."""

from __future__ import annotations

import re
from datetime import date

from fastapi import HTTPException, status

from app.models.question import Question
from app.services.geocoding import mapbox_configured, search_places

MAX_TRIP_NIGHTS = 14
MAX_TRAVELERS_TOTAL = 20
MIN_CITY_LENGTH = 2
MAX_CITY_LENGTH = 120

CITY_PATTERN = re.compile(r"^[\w\s,'.\-()]+$", re.UNICODE)


def _city_label_matches(query: str, label: str) -> bool:
    q = query.strip().lower()
    lab = label.strip().lower()
    return q == lab or q in lab or lab in q


def validate_city_name(value: str, *, field: str) -> str:
    cleaned = value.strip()
    if len(cleaned) < MIN_CITY_LENGTH:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=f"{field} must be at least {MIN_CITY_LENGTH} characters.",
        )
    if len(cleaned) > MAX_CITY_LENGTH:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=f"{field} must be at most {MAX_CITY_LENGTH} characters.",
        )
    if not CITY_PATTERN.match(cleaned):
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=f"{field} contains invalid characters.",
        )

    if mapbox_configured():
        results = search_places(cleaned, limit=8)
        if results and not any(_city_label_matches(cleaned, r["label"]) for r in results):
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
                detail=f"Could not verify {field.lower()}. Pick a city from search suggestions.",
            )

    return cleaned


def validate_date_range(value: dict) -> tuple[date, date]:
    start_raw = value.get("start")
    end_raw = value.get("end")
    if not start_raw or not end_raw:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="Select both departure and return dates.",
        )

    try:
        start = date.fromisoformat(start_raw) if isinstance(start_raw, str) else start_raw
        end = date.fromisoformat(end_raw) if isinstance(end_raw, str) else end_raw
    except (TypeError, ValueError) as exc:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="Invalid date format.",
        ) from exc

    today = date.today()
    if start < today:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="Departure cannot be in the past.",
        )
    if end < start:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="Return date must be on or after departure.",
        )

    nights = (end - start).days
    if nights > MAX_TRIP_NIGHTS:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=f"Trips can be up to {MAX_TRIP_NIGHTS} nights.",
        )

    return start, end


def validate_travelers(value: dict) -> tuple[int, int]:
    try:
        adults = int(value.get("adults", 1))
        children = int(value.get("children", 0))
    except (TypeError, ValueError) as exc:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="Invalid traveler counts.",
        ) from exc

    if adults < 1:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="At least one adult is required.")
    if children < 0:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Children count cannot be negative.")
    if adults > 20 or children > 10 or adults + children > MAX_TRAVELERS_TOTAL:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=f"Maximum {MAX_TRAVELERS_TOTAL} travelers per trip.",
        )
    return adults, children


def validate_choice(question: Question, value) -> str | list[str]:
    allowed = {opt.option_key for opt in question.options}

    if question.multi:
        if not isinstance(value, list) or not value:
            raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Select at least one option.")
        if any(v not in allowed for v in value):
            raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Invalid selection.")
        return value

    if not isinstance(value, str) or value not in allowed:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Please select a valid option.")
    return value


def validate_answer(question: Question, value) -> None:
    if question.question_type == "choice":
        validate_choice(question, value)
        return

    if question.question_type == "date_range":
        if not isinstance(value, dict):
            raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Invalid date range.")
        validate_date_range(value)
        return

    if question.question_type == "travelers":
        if not isinstance(value, dict):
            raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Invalid travelers value.")
        validate_travelers(value)
        return

    if question.question_type == "text":
        if not isinstance(value, str):
            raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Invalid text answer.")
        if question.key in ("origin", "destination"):
            validate_city_name(value, field="Origin" if question.key == "origin" else "Destination")
        elif not value.strip():
            raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Answer cannot be empty.")


def validate_quiz_submission(
    phase: str,
    question_map: dict[int, Question],
    answers_by_id: dict[int, object],
) -> None:
    answers_by_key: dict[str, object] = {}
    for question_id, value in answers_by_id.items():
        question = question_map.get(question_id)
        if question is None:
            continue
        answers_by_key[question.key] = value
        validate_answer(question, value)

    if phase == "quiz":
        required = {"destination_known", "dates", "travelers", "origin"}
        if answers_by_key.get("destination_known") != "not_sure":
            required.add("destination")

        missing = required - set(answers_by_key.keys())
        if missing:
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
                detail="Please complete all quiz steps before continuing.",
            )

        origin = answers_by_key.get("origin")
        destination = answers_by_key.get("destination")
        if (
            isinstance(origin, str)
            and isinstance(destination, str)
            and origin.strip().lower() == destination.strip().lower()
        ):
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
                detail="Origin and destination must be different cities.",
            )

    elif phase == "preferences":
        required = {"trip_purpose", "theme", "budget_tier", "include_flights", "include_hotels"}
        missing = required - set(answers_by_key.keys())
        if missing:
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
                detail="Please complete all preference steps.",
            )
